### PR TITLE
Fix: Support the List within List type in withParam #1471

### LINKED ISSUE #1471: Support the List within List type in withParam
**Is this a BUG REPORT or FEATURE REQUEST?**:
 FEATURE REQUEST
**What happened**:
`  - name: test-java-wf-example
    steps:
    - - name: generate
        template: gen-name-list
    - - name: print
        template: print-name
        arguments:
          parameters:
          - name: seconds
            value: "{{item}}"
        withParam: "{{steps.generate.outputs.result}}"`
Generate step will output list contains list. Print step need to take array as argument.
Received Error.
`STEP                   PODNAME                        DURATION  MESSAGE
 ✖ test-java-wf-5khgf                                           withItems[0] expected string, number, or map. received: [Kooh0uSs7I lBuIdrV5jX l9eN5sfPJM]
 ├---✔ generate        test-java-wf-5khgf-2805467061  7s        
`

### PR DESCRIPTION
**Pull request by sarabala1979**
This PR will fix #1471

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -1637,6 +1637,13 @@ func processItem(fstTmpl *fasttemplate.Template, name string, index int, item wf
 		// sort the values so that the name is deterministic
 		sort.Strings(vals)
 		newName = fmt.Sprintf("%s(%d:%v)", name, index, strings.Join(vals, ","))
+	case []interface{}:
```
Comment (xianlubird) [inline workflow/controller/operator.go]: Maybe we should add a test yaml to show this function.

**Comment by jessesuen** [comment]
Can we add a unit test for this?

File: workflow/controller/operator_test.go [hunk 2]
```diff
@@ -1159,3 +1159,50 @@ func TestResourceWithOwnerReferenceTemplate(t *testing.T) {
...
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  name: expand-with-items
+spec:
+  entrypoint: expand-with-items
+  arguments:
+    parameters:
+    - name: input
+      value: '[1,2,3,4]'
```
Comment (jessesuen) [inline workflow/controller/operator_test.go]: How is this test testing list of lists? I don't think your unit test is actually testing the feature.

Comment (sarabala1979) [reply workflow/controller/operator_test.go]: This fix is mainly if WithParam has a json array. The controller needs to execute a step for each item in the array. If WithParam has a list of lists,  Controller will execute the referred step for main array items.
Eg: withParam: [[1,2],[3,4],[5,6]]
the step will execute for 3 times with each item and pass sub-array as argument

Comment (jessesuen) [reply workflow/controller/operator_test.go]: Yea, but your `withParam` is not a list of lists. It's: `[1,2,3,4]`, which is a single level.

File: examples/loops-param-result-array.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# This workflow demonstrates the use of a generator step which produces a list within list of items as a result.
```
Comment (jessesuen) [inline examples/loops-param-result-array.yaml]: Can we make this a test and not an example. Don't think it's necessary to make as an example

File: workflow/controller/operator.go [hunk 4]
```diff
@@ -1637,6 +1637,13 @@ func processItem(fstTmpl *fasttemplate.Template, name string, index int, item wf
...
+		byteVal, err := json.Marshal(val)
+		if err != nil {
+			return "", errors.InternalWrapError(err)
+		}
+		replaceMap["item"] = string(byteVal)
+		newName = fmt.Sprintf("%s(%d:%v)", name, index, val)
 	default:
 		return "", errors.Errorf(errors.CodeBadRequest, "withItems[%d] expected string, number, or map. received: %s", index, val)
```
Comment (jessesuen) [inline workflow/controller/operator.go]: This message needs to be updated.

**Pull request merged**